### PR TITLE
Object GC for block splitting inside the dataset splitting

### PR DESCRIPTION
**Pull request by jianoaix**
## Why are these changes needed?
The pipeline will spill objects when splitting the dataset into multiple equal parts.

## Related issue number
#25249

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -3400,6 +3400,11 @@ def _split(
                 left_metadata.append(ray.get(m0))
                 right_blocks.append(b1)
                 right_metadata.append(ray.get(m1))
+                # If return_right_half is requested, the input block b will be copied
+                # into b0 and b1. In such case, we can safely clear b if this is in
+                # lazy mode.
```
Comment (ericl) [inline python/ray/data/dataset.py:3405]: Hmm isn't it copied in both cases?

Comment (jianoaix) [reply python/ray/data/dataset.py:3405]: At the level of this method, the right half may not copied and conditioned on this boolean: https://sourcegraph.com/github.com/ray-project/ray@master/-/blob/python/ray/data/dataset.py?L3689

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -3400,6 +3400,11 @@ def _split(
...
+                if return_right_half and self._lazy:
+                    ray._private.internal_api.free(b, local_only=False)
```
[the remark this answers is not in the document]
Comment (jianoaix) [reply python/ray/data/dataset.py:3407]: Yes, we should do it more cleanly. We may need to have an overhaul of execution semantics, e.g. 1) we execute the plan when it's split(), regardless of whether it's lazy or not; 2) we need to have the clear semantics of eager v.s. cached/pinned dataset (via running ds.fully_executed()), the later will make the dataset like an eager one, but running fully_executed() isn't going to make a lazy dataset eager.

File: python/ray/data/dataset.py (hunk 1 above)
Comment (ericl) [reply python/ray/data/dataset.py:3405]: But the left half is always copied. And the right half is either copied or None. It seems good to clear in either case.

**Review by clarkzinzow — COMMENTED** [review]
Since this is force-clearing on the base blocks of the dataset, we should be careful that this won't break certain fan-out patterns. E.g.:
```python
ds = ray.data.from_items(list(range(1000))).experimental_lazy()
ds.limit(10).show()
ds.show()

ds2 = ray.data.from_items(list(range(1000))).experimental_lazy()
dses = ds.split_at_indices([250, 750])
for ds in dses:
    ds.show()
ds2.show()
```

The first should work since we're only clearing if `return_right_half=True` (although agreed with @ericl that that shouldn't matter), but I think the second might fail.

**Comment by jianoaix** [comment]
I looked more into this regarding safety. It can happen actually quite easily (not even need to fanout or use from_items()). For example, this will crash:
```
ds1 = ray.data.range(100, parallelism=10).map_batches(lambda x: x).experimental_lazy() # map_batches() can be dropped
ds1_splits = split(3, equal=True)
for ds in ds1_splits:
    for batch in ds.iter_batches():
        pass
for batch in ds1.iter_batches():
    pass
```
It's generally more tricky to do eager object gc for dataset than pipeline, as the later is protected by the API semantics that it can be consumed once at most.

We might still be able to do eager gc for lazy dataset, if we move to a semantics like "each time to execute a lazy dataset, it will be executed from the beginning" ("beginning" would mean, for example, the source input blocks, or some intermediate dataset which is known to be pinned and cannot be cleared before unpinned, the output of eager dataset). This will need a revamp work (filed https://github.com/ray-project/ray/issues/26264) first (at least part of it).

One hack we may try is to flag the Dataset (with an extra parameter) whether it's used in DatasetPipeline processing, and if so we clear the input block after split.

**Comment by jianoaix** [comment]
@ericl @clarkzinzow This should be ready to review.

**Review by clarkzinzow — APPROVED** [review]
LGTM!

**Pull request merged**